[clang][analyzer] Improve BlockInCriticalSectionsChecker

gamesh411 opened the pull request:
On entering a critical section, a note tag is now placed along the bugpath.

gamesh411 committed:
[clang][analyzer] Improve BlockInCriticalSectionsChecker with multi-section and recursive mutex support

* Add support for multiple, potentially overlapping critical sections:
  The checker can now simultaneously handle several mutex's critical
  sections without confusing them.
* Implement the handling of recursive mutexes:
  By identifying the lock events, recursive mutexes are now supported.
  A lock event is a pair of a lock expression and the SVal of the mutex
  that it locks, so even multiple locks of the same mutex (and even by
  the same expression) is now supported.
* Refine the note tags generated by the checker:
  The note tags now correctly show just for mutexes those are
  active at point of error, and multiple acqisitions of the same mutex
  are also noted.

NagyDonat reviewed (approved):
LGTM.

balazske commented:
I think we could get a note at the first `lock` call in this code, this behavior is not entirely correct:
```
void test() {
  std::mutex m;
  m.lock();
  m.unlock();
  m.lock();
  sleep(3);
  m.unlock();
}
```

balazske commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -57,6 +61,12 @@ class BlockInCriticalSectionChecker : public Checker<check::PostCall> {
+                                             CheckerContext &C) const;
+  const NoteTag *
+  createCriticalSectionNote(const CritSectionMarker &CriticalSectionBegin,
+                            CheckerContext &C) const;
```
I do not know if it is sufficient to store the set of all found lock calls until the last unlock happens. Probably in the following situation it does not work:
```
std::mutex m, n, k;
m.lock();
n.lock();
n.unlock();
k.lock();
k.unlock();
sleep(1);
m.unlock();
```
It looks better to store all locked mutex objects somehow (not the lock calls), and remove one at unlock. Then the mutex counter is not needed.

gamesh411 commented:
As keeping the trivial change that this PR was before, and thanks to the fact that the checker needed a major overhaul, I have increased the scope of this change to support the use cases mentioned by @balazske.
That is why I am re-requesting a review.

NagyDonat reviewed (commented):
Seems to be a good change overall and I like that you implemented this with some nice helper classes. However, there is some code duplication that could be reduced (see inline comments).

Also, I feel that (especially after these extensions) this checker duplicates some logic that also appears in `PthreadLockChecker.cpp`. I didn't check the exact details but that checker is also modeling the locking and unlocking calls in order to report bad locking/unlocking order.

In theory it would be to ensure that we don't develop and maintain two similar but different solutions that both track lock handling code -- but PthreadLockChecker is 700 lines long and I understand that it would be difficult to unify it with this checker.

gamesh411 committed:
reduce code duplication

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -70,73 +202,121 @@ class BlockInCriticalSectionChecker : public Checker<check::PostCall> {
+      });
+  if (UnlockDescriptor != MutexDescriptors.end())
+    return *UnlockDescriptor;
+  return std::nullopt;
```
I think I have fixed most of the code duplication, thanks for the suggestions 👍

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -20,48 +20,180 @@
+    auto *IdentifierInfo =
+        cast<CXXRecordDecl>(Dtor->getDecl()->getParent())->getIdentifier();
+    return IdentifierInfo == Guard;
+  }
```
Fixed.

gamesh411 commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -20,48 +20,180 @@
+  [[nodiscard]] const MemRegion *getUnlockRegion(const CallEvent &Call) const {
+    return cast<CXXMemberCall>(Call).getCXXThisVal().getAsRegion();
+  }
+};
```
Introduced a common base class for these 2 classes.

gamesh411 committed:
Use type deduction

MSVC has an implementation where the return value of std::find_if is a proper iterator not a pointer
In order to support that, we use type deduction now

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -20,48 +20,178 @@
+                              CallDescription &&UnlockFn)
+      : LockFn(std::move(LockFn)), UnlockFn(std::move(UnlockFn)) {}
+  [[nodiscard]] bool matches(const CallEvent &Call, bool IsLock) const {
+    if (IsLock) {
+      return LockFn.matches(Call);
+    }
```
```suggestion
    if (IsLock)
      return LockFn.matches(Call);
```
Bikeshedding: the coding standard says that braces shouldn't be used on simple single-statement bodies: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Unorthodox but (IMO) elegant alternative idea: replace the whole body of the function with
```
return (isLock ? LockFn : UnlockFn).matches(Call);
```

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -20,48 +20,178 @@
+  RAIIMutexDescriptor(StringRef GuardName) : GuardName(GuardName) {}
+  [[nodiscard]] bool matches(const CallEvent &Call, bool IsLock) const {
+    initIdentifierInfo(Call);
+    if (IsLock) {
+      return matchesImpl<CXXConstructorCall>(Call);
+    }
```
```suggestion
    if (IsLock)
      return matchesImpl<CXXConstructorCall>(Call);
```
Bikeshedding as above.

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/BlockInCriticalSectionChecker.cpp:
```diff
@@ -20,48 +20,178 @@
+                                           bool IsLock) const {
+    const MemRegion *LockRegion = nullptr;
+    if (IsLock) {
+      if (std::optional<SVal> Object = Call.getReturnValueUnderConstruction()) {
+        LockRegion = Object->getAsRegion();
+      }
```
```suggestion
      if (std::optional<SVal> Object = Call.getReturnValueUnderConstruction())
        LockRegion = Object->getAsRegion();
```
Yet another brace tweak...

NagyDonat reviewed (approved):
Thanks for simplifying the code :)

I added some stereotypical bikeshedding in inline comments, but apart from that the commit LGTM.

balazske commented:
It looks like that this change causes crashes on many projects (curl, vim, postgres, others) in `RAIIMutexDescriptor::initIdentifierInfo`.
